kong/conf_loader.lua

```diff
@@ -560,6 +565,47 @@ local function check_and_infer(conf)
     errors[#errors + 1] = "router_update_frequency must be greater than 0"
   end
 
+  if conf.role == "control_plane" then
+    if #conf.admin_listen < 1 or pl_stringx.strip(conf.admin_listen[1]) == "off" then
+      errors[#errors + 1] = "admin_listen must be specified when role = \"control_plane\""
+    end
+
+    if #conf.cluster_listen < 1 or pl_stringx.strip(conf.cluster_listen[1]) == "off" then
+      errors[#errors + 1] = "cluster_listen must be specified when role = \"control_plane\""
+    end
+
+    if conf.database == "off" then
+      errors[#errors + 1] = "in-memory storage can not be used when role = \"control_plane\""
+    end
+
+  elseif conf.role == "data_plane" then
+    if #conf.proxy_listen < 1 or pl_stringx.strip(conf.proxy_listen[1]) == "off" then
+      errors[#errors + 1] = "proxy_listen must be specified when role = \"data_plane\""
+    end
+
+    if conf.database ~= "off" then
+      errors[#errors + 1] = "only in-memory storage can be used when role = \"data_plane\"\n" ..
+                            "Hint: set database = off in your kong.conf"
+    end
+  end
+
+  if conf.role == "control_plane" or conf.role == "data_plane" then
+    if not conf.cluster_cert or not conf.cluster_cert_key then
```

Comment:
Is there any practical reason why `control-plane` couldn't just generate them when missing? Also do we really share same private key and public key with both control plane and dataplane? That feels rather messy? Can't we derive data-plane certs from control plane cert so that we never copy control plane private keys to dataplanes.